RFC - not for merging - alternative uniform handling

aardgoose opened the pull request:
I've been looking at #8552, #9927 and the issues around fixing instances / interactive / gpu single material example.

This is a prototype of the program pulling uniforms from the renderer rather than hard coding the requirements in webGLRenderer setProgram().

I've kept it opt in  ( Material.isExperimentalMaterial = true ) to allow easy testing without breaking existing code. The new uniform values are called parameters in the code to prevent confusion with existing uniform handling code.

Features:

- ShaderMaterial uniforms can be handled as normal material properties such as color, opacity etc. myMaterial.parameter = x; rather than myMaterial.uniforms.parameter.value = x;

- Parameters can be pulled from a hierarchy of sources ( implemented object -> material in prototype).
- Parameters are versioned, no needsUpdate required at object/material/parameter.
( changes to object values eg THREE.Color() are also handled by updating a parameter version on get as well as set operations). versioning is transparent to the user.
- Only changed uniforms are sent to the GPU.
- If extended to built in materials could remove a lot of the custom per material uniform setting in setProgram.

Obviously this is very raw and things like cloning/copying parameters are not implemented.

Is this worth exploring further, say converting some built in materials to this method?
Any interest in this proposal, comments? 

This PR also includes fixing the instances / interactive /gpu example which removes the need for the removed dynamic Uniform or onBeforeRender() callbacks (which don't work for this case anyway) by using object specific parameters.

aardgoose committed:
start test setProgram method

aardgoose committed:
WIP add MaterialParameter object

aardgoose committed:
prototype parameter handling

aardgoose committed:
invalidate parameter cache on material change

aardgoose committed:
set initial parameter versions to be unique

aardgoose committed:
use experimental material to fix example.

aardgoose committed:
minor fix

aardgoose committed:
remove noise from old version

aardgoose committed:
remove noise

aardgoose committed:
missing declaration

aardgoose committed:
cache per object/material pair & remove chrom deopts

fallenoak commented:
I'll try give this a spin later this weekend.

From your description, it sounds like a pretty comprehensive set of improvements to the issues with the existing uniform system.

aardgoose committed:
minor refactor and cleanup

aardgoose committed:
add basic camera parameters

aardgoose committed:
add object3D matrixWorld parameter with alias

aardgoose committed:
add mesh basic parameters

aardgoose committed:
attach version cache to program and use camera parameters

aardgoose committed:
add support for LineBasicMaterial

aardgoose committed:
fix quotes

aardgoose committed:
fix quotes

aardgoose committed:
add Texture offsetRepeat support

aardgoose committed:
convert MeshLambertMaterial

aardgoose committed:
add renderer param source and work around for texture issue

aardgoose committed:
fix Texture offset/repeat parameters

aardgoose committed:
add FIXME comment

aardgoose committed:
Merge branch 'dev' into parameters

aardgoose committed:
update pahysical and standard materials

aardgoose committed:
fix issue with image loading and shaderlib uniforms

aardgoose commented:
OK, a I counldn't resist.

LineBasic, MeshBasic, MeshLambert, MeshStandard & MeshPhysical materials converted over and working for most examples.  MeshPhong and PointsMaterial require mechanism to adjust parameters before setting uniforms. Fog needs putting in here to.

Existing code path can be easily restored by removing the .isExperimentalMaterial flag.

mrdoob commented on src/materials/MeshLambertMaterial.js:
```diff
@@ -46,28 +46,30 @@ function MeshLambertMaterial( parameters ) {
+	this.isExperimentalMaterial = true;
 
-	this.map = null;
+	this.addParameter( 'color', new Color( 0xffffff ), 'diffuse' ); // diffuse
```
Maybe a less intrusive approach could be something like... `this.defineParameters( this, [ 'color', 'lightMap', 'lightMapIntensity', ... ] );

aardgoose committed:
fog and log depth buffer support

aardgoose committed:
log depth buffer and fog support II

aardgoose commented:
Could do, however some names are aliased eg color -> diffuse. As an alternative it could parse through UniformsLib.<shader>.uniforms and add an .alias.  and handle ShaderMaterials in a similar way, almost making built in and custom shaders modular (apart from things like DEFINES which are another issue). That could remove a lot of code from built in materials and stop the duplication in Material objects and UniformsLibs.  

Is this whole approach interesting? It certainly makes uniforms a lot more dynamic (overrides by Object etc) and creates a much simplified setProgram, and avoids all the "material.uniform.<name>.value".

mrdoob commented:
> Is this whole approach interesting? It certainly makes uniforms a lot more dynamic (overrides by Object etc) and creates a much simplified setProgram, and avoids all the "material.uniform..value".

Yeah, the current approach clearly needs some re-architecting. Keep in mind serialisation and the structure you're creating when doing so though.

aardgoose committed:
make value versions - value for plain numbers

aardgoose committed:
fix shadow material ( inherits opacity form Material

aardgoose committed:
add parameters to Skeleton and SkinnedMesh

aardgoose committed:
WIP improve state tracking

aardgoose committed:
support LineDashedMaterial

aardgoose committed:
test version of ShaderMaterial

aardgoose committed:
code move to contain within setProgram*

aardgoose committed:
add version and isSingleUniform to simplify renderer

aardgoose committed:
remove level of indirection and uneeded version cache

aardgoose committed:
more tidyup of redundant code

aardgoose committed:
convert toon material

aardgoose committed:
convert remaining mesh materials, alias emissive

aardgoose committed:
adapt points material

aardgoose committed:
simplify tagging of uniform objects

aardgoose committed:
fix camera position

aardgoose committed:
Merge branch 'dev' into parameters

aardgoose committed:
cleanup

aardgoose committed:
add DerivedMaterialParameter

aardgoose committed:
use derived parameter to limit range of shininess

aardgoose committed:
handle camera changes efficiently and rework cache

aardgoose committed:
Merge branch 'dev' into parameters

aardgoose committed:
rename parameter

aardgoose committed:
move flipEnvMap to materials

aardgoose committed:
set flipEnvMap in materials

aardgoose committed:
split parameter lists

aardgoose committed:
revert to flipEncMap setting

aardgoose committed:
remove debug code

aardgoose committed:
optimisations

aardgoose committed:
major simplification

aardgoose committed:
Merge branch 'dev' into parameters

aardgoose committed:
remove WS damage

aardgoose committed:
remove blank line

aardgoose committed:
remove unused var

aardgoose committed:
more WS cleanup

aardgoose committed:
more WS cleanup

aardgoose committed:
pu boolean on prototype

aardgoose committed:
fix breakages from cleanup and remove workaround for emissive from core.

aardgoose committed:
cleanup line materials

aardgoose committed:
add missing emissive o StandardMaterial

aardgoose commented:
Ok, this is more sound I think,  dropped versioning - too much js performance overhead, and also simplified things a lot.  Nearly examples work with no change with built in materials, RectArea lighting doesn't work because the pre-computed textures are not built in yet. Very little material specific knowledge in the renderer required now.

aardgoose committed:
remove the need for closures in callbacks

aardgoose committed:
fixup offsetRepeat properly

aardgoose committed:
revert original code

aardgoose committed:
fix WS

aardgoose committed:
complete merge

aardgoose committed:
remove cut/paste error

aardgoose committed:
Merge branch 'dev' into parameters

arose commented:
Is this ready for some testing? I would really like to see something like this included so that dynamic updating of uniforms can be done again.

aardgoose committed:
fix merge

aardgoose commented:
minor merge fix to keep in line with dev. 
AFAICS all examples work correctly.

As code it works. The interactive / instances / gpu examples demonstrates per object uniforms but the onBeforeRender handler could equally be used. Or even addDerivedParameter() gives you something equivalent to the old dynamic uniform.
 
I haven't looked at light uniforms, I think any work there should take into account the possibility of using UBOs in WebGL2 for lighting uniforms.  I'd certainly be interested in any results.

aardgoose committed:
Merge branch 'dev' into parameters

aardgoose committed:
fix merge

aardgoose committed:
Merge branch 'dev' into parameters

aardgoose commented:
Ok, latest merge conflicts fixed.

@mrdoob any thoughts?

aardgoose committed:
Merge branch 'dev' into parameters

aardgoose commented:
Dropping this one, much simplified version to come,
